### PR TITLE
Test functionality

### PR DESCRIPTION
**Pull request by pmisteliac**
* Added Test for local network with 2 actors
* Reworked logging
* Fixed bugs in one-hop replication
* Made messages serializable
* Numerous refactorings

### CODE REVIEW

File: src/main/scala/GiaActor.scala [hunk 1]
```diff
@@ -154,60 +123,57 @@ class GiaActor(ip: String, port: Int) extends Actor with ActorLogging with Timer
...
   }
 
-  private def searchProcedure(msg : Search) : Unit = {
-    // Logging search start.
+  private def searchFile(msg : Search) : Unit = {
     log.info("Searching for file: " + msg.keyword +" with time to live of " + msg.timeToLive)
 
-    // Check if message has not already been received/processed.
-    if(messageBufferObj.containsMessage(msg.uniqueID)) {
+    if (msg.timeToLive <= 0) {
```
Comment (ruteben) [inline src/main/scala/GiaActor.scala]: I understand that we should check the time to live, but shouldn't we also notify the original sender of this?

File: src/main/scala/GiaActor.scala [hunk 2]
```diff
@@ -154,60 +123,57 @@ class GiaActor(ip: String, port: Int) extends Actor with ActorLogging with Timer
...
 
     // Priority queue, ordered by node capacity largest -> smallest, of all neighbors with the file.
-    var fileContainedInNodes = new mutable.ListBuffer[GiaNode]
+    var neighborsWithFile = new mutable.ListBuffer[GiaNode]
 
     // Checking if the current actor has the file. If so, enqueue and broadcast.
     if(localContent.contains(msg.keyword)) {
       msg.origin ! Found(localContent.get(msg.keyword))
+      log.info("File found locally: " + msg.keyword)
       return
```
Comment (ruteben) [inline src/main/scala/GiaActor.scala]: I'm still not entirely convinced that the current node should always be the one to send the file, especially if there could be neighbours that have a greater capacity that also have the file. Could someone please explain why the current approach is the correct one?

**Comment by ruteben** [comment]
Shouldn't we make use of the akka testkit? https://doc.akka.io/docs/akka/2.5/testing.html
That way we could use assertions, check whether certain log messages are written, we can even override the scheduler and test those timed features.
Thoughts?

**Comment by pmisteliac** [comment]
I'll look into it. But it definitely has to work outside the test framework as well. I don't think having a test suite was a requirement for the project.

**Comment by dalderliesten** [comment]
Looks fine to me.

**Pull request merged**